### PR TITLE
Cleaning unittests

### PR DESCRIPTION
**Pull request by rekonder**
Changes:
- Adding iterators in tests where is similar code, but only parameters are changing
- Deleting some duplicated rows
- Adding some functions into setUpClass and changed some assertTrue into assertGreater...

### CODE REVIEW

File: Orange/tests/test__orange.py [hunk 1]
```diff
@@ -4,38 +4,26 @@
...
-
-        a = np.array([[0, 1, 2, 3], [0.1, 0.2, 0.3, 0.4]])
-        b = _valuecount.valuecount(a)
-        np.testing.assert_almost_equal(b, a)
-
-        a = np.array([[0], [0.1]])
-        b = _valuecount.valuecount(a)
-        np.testing.assert_almost_equal(b, a)
+    def valuecount_helper(self, *args):
+        for arg in args:
```
Comment (kernc) [inline Orange/tests/test__orange.py]: Prefer unpacking to `args` indexing:
```py
for a, expected_b in args:
```

File: Orange/tests/test__orange.py [hunk 2]
```diff
@@ -4,38 +4,26 @@
...
-        np.testing.assert_almost_equal(b, [[0, 1, 2], [0.1, 0.5, 0.4]])
-
-        a = np.array([[0, 1, 2, 3], [0.1, 0.2, 0.3, 0.4]])
-        b = _valuecount.valuecount(a)
-        np.testing.assert_almost_equal(b, a)
-
-        a = np.array([[0], [0.1]])
-        b = _valuecount.valuecount(a)
-        np.testing.assert_almost_equal(b, a)
+    def valuecount_helper(self, *args):
```
Comment (kernc) [inline Orange/tests/test__orange.py]: Perhaps there's no need for this to be a separate function, and you can just move its body to where it's used?

File: Orange/tests/test_color.py [hunk 3]
```diff
@@ -9,21 +9,19 @@
...
+
     def test_two_color(self):
         generator = GradientPaletteGenerator('#000', '#fff')
-        np.testing.assert_equal(generator.getRGB(.5), (128, 128, 128))
-        np.testing.assert_equal(generator.getRGB(np.nan), NAN_GREY)
-        np.testing.assert_equal(generator.getRGB([0, .5, np.nan]),
-                                [(0, 0, 0),
-                                 (128, 128, 128),
-                                 NAN_GREY])
+        for arg in ((.5, (128, 128, 128)),
```
Comment (kernc) [inline Orange/tests/test_color.py]: Instead of `arg[0]` and `arg[1]` below, use:
```py
for float_values, rgb in ...
```
Please perform this kind of rationalized unpacking everywhere where applicable.

File: Orange/tests/test_color.py [hunk 4]
```diff
@@ -9,21 +9,19 @@
...
     def test_three_color(self):
         generator = GradientPaletteGenerator('#f00', '#000', '#fff')
-        np.testing.assert_equal(generator.getRGB(.5), (0, 0, 0))
-        np.testing.assert_equal(generator.getRGB(np.nan), NAN_GREY)
-        np.testing.assert_equal(generator.getRGB([0, .5, 1]),
-                                [(255, 0, 0),
-                                 (0, 0, 0),
-                                 (255, 255, 255)])
-
+        for arg in ((.5, (0, 0, 0)), (np.nan, NAN_GREY), ((0, .5, 1), ((255, 0, 0),
```
Comment (kernc) [inline Orange/tests/test_color.py]: Prefer one param group per line as you did above.

File: Orange/tests/test_datasets.py [hunk 5]
```diff
@@ -5,22 +5,23 @@
...
+        fname = datasets.anneal['location']
+        d2 = datasets['anneal']
         self.assertNotEqual(len(d1), 0)
         self.assertEqual(len(d1), len(d2))
 
     def test_filter(self):
-        discr = [info for info in Orange.datasets.values()
+        discr = [info for info in datasets.values()
                  if info['features']['continuous'] == 0]
         for info in discr:
```
Comment (kernc) [inline Orange/tests/test_datasets.py]: Since `discr` isn't used anywhere else, you could easily merge these two iterations into one (this) `for` and one `if` beneath it.

File: Orange/tests/test_distribution.py [hunk 6]
```diff
@@ -354,16 +341,15 @@ def test_get_distributions(self):
         np.testing.assert_almost_equal(ddist[2], freqs)
         np.testing.assert_almost_equal(ddist[-1], [50, 50, 50])
 
-
-    #noinspection PyTypeChecker
+    # noinspection PyTypeChecker
```
Comment (kernc) [inline Orange/tests/test_distribution.py]: What does this do?

File: Orange/tests/test_domain.py [hunk 7]
```diff
@@ -124,20 +125,11 @@ def test_init_metas(self):
...
-        d = Domain.from_numpy(np.zeros((1, 99)))
-        self.assertTrue(d.anonymous)
-        self.assertEqual([var.name for var in d.attributes],
-                         ["Feature {:02}".format(i) for i in range(1, 100)])
-
-        d = Domain.from_numpy(np.zeros((1, 100)))
-        self.assertTrue(d.anonymous)
-        self.assertEqual([var.name for var in d.attributes],
-                         ["Feature {:03}".format(i) for i in range(1, 101)])
+        for x, y in [(5, "Feature {}"), (99, "Feature {:02}"), (100, "Feature {:03}")]:
```
Comment (kernc) [inline Orange/tests/test_domain.py]:
```py
for n_cols, name in ...
```

Comment (kernc) [inline Orange/tests/test_domain.py]: + one param group per line.

File: Orange/tests/test_domain.py [hunk 8]
```diff
@@ -153,41 +145,32 @@ def test_from_numpy_names(self):
...
-        d = Domain.from_numpy(np.zeros((1, 1)), np.arange(2).reshape(2, 1))
-        self.assertTrue(d.anonymous)
-        self.assertIsInstance(d.class_var, DiscreteVariable)
-        self.assertEqual(d.class_var.values, ["v{}".format(i)
-                                              for i in range(1, 3)])
-
-        d = Domain.from_numpy(np.zeros((1, 1)), np.arange(18, 23).reshape(5, 1))
-        self.assertTrue(d.anonymous)
-        self.assertIsInstance(d.class_var, ContinuousVariable)
+        for aran_min, aran_max, res_min, res_max, type in [(1, 3, 2, 1, ContinuousVariable),
```
Comment (kernc) [inline Orange/tests/test_domain.py]: * `res_min` and `res_max` define shape (i.e. n_rows, n_cols); nothing to do with min/max.
* All `res_min` are equal to `aran_max - aran_min`, therefore redundant. All `res_max` are `1`.
* Don't occlude a builtin function by going with, e.g., `vartype`.

File: Orange/tests/test_domain.py [hunk 9]
```diff
@@ -210,39 +193,19 @@ def test_wrong_types(self):
...
-        self.assertEqual(d[income], income)
-        self.assertEqual(d["income"], income)
-        self.assertEqual(d[2], income)
-
-        self.assertEqual(d[ssn], ssn)
-        self.assertEqual(d["SSN"], ssn)
-        self.assertEqual(d[-1], ssn)
-
-        self.assertEqual(d[-2], race)
+        for x, y in [(age, age), ("AGE", age), (0, age),
```
Comment (kernc) [inline Orange/tests/test_domain.py]:
```py
for idx, var in ...
```
\+ one param group per line.

File: Orange/tests/test_domain.py [hunk 10]
```diff
@@ -210,39 +193,19 @@ def test_wrong_types(self):
...
-        self.assertEqual(d.index(np.int_(2)), 2)
-
-        self.assertEqual(d.index(ssn), -1)
-        self.assertEqual(d.index("SSN"), -1)
-        self.assertEqual(d.index(-1), -1)
-        self.assertEqual(d.index(np.int_(-1)), -1)
-
-        self.assertEqual(d.index(-2), -2)
-        self.assertEqual(d.index(np.int_(-2)), -2)
+        for x, y in [(age, 0), ("AGE", 0), (0, 0), (np.int_(0), 0),
```
Comment (kernc) [inline Orange/tests/test_domain.py]: As above.

File: Orange/tests/test_domain.py [hunk 11]
```diff
@@ -252,55 +215,29 @@ def test_get_item_slices(self):
...
         d = Domain((age, gender, income), metas=(ssn, race))
-        with self.assertRaises(KeyError):
-            _ = d[3]
-        with self.assertRaises(KeyError):
-            _ = d[-3]
-        with self.assertRaises(KeyError):
-            _ = d[incomeA]
-        with self.assertRaises(KeyError):
-            _ = d["no_such_thing"]
+        for x in (3, -3, incomeA, "no_such_thing"):
```
Comment (kernc) [inline Orange/tests/test_domain.py]:
```py
for idx in ...
```

File: Orange/tests/test_domain.py [hunk 12]
```diff
@@ -252,55 +215,29 @@ def test_get_item_slices(self):
...
-            d.index(np.int_(3))
-        with self.assertRaises(ValueError):
-            d.index(-3)
-        with self.assertRaises(ValueError):
-            d.index(np.int_(-3))
-        with self.assertRaises(ValueError):
-            d.index(incomeA)
-        with self.assertRaises(ValueError):
-            d.index("no_such_thing")
+        for x in (3, np.int(3), -3, np.int(-3), incomeA, "no_such_thing"):
```
Comment (kernc) [inline Orange/tests/test_domain.py]: `idx`, or `var`, or something less generic.

File: Orange/tests/test_domain.py [hunk 13]
```diff
@@ -373,55 +310,24 @@ def test_has_continuous(self):
         self.assertTrue(Domain([], [race, age]).has_continuous_attributes(True))
 
     def test_get_conversion(self):
+        x = lambda: 42
+        new_income = income.copy(compute_value=x)
```
Comment (kernc) [inline Orange/tests/test_domain.py]: Why store `x`?

File: Orange/tests/test_domain.py [hunk 14]
```diff
@@ -505,20 +411,19 @@ def test_domain_conversion_is_fast_enough(self):
...
-        self.assertEqual(c3.class_vars, list(range(-1, -11, -1)))
-        self.assertEqual(c3.metas, list(range(1000)))
+        cases = (((attrs[:1000], class_vars, metas),
+                  list(range(1000)), list(range(10000, 10010)), list(range(-1, -11, -1))),
+                 ((metas, attrs[:1000], class_vars),
+                  list(range(-1, -11, -1)), list(range(1000)), list(range(10000, 10010))),
+                 ((class_vars, metas, attrs[:1000]),
+                  list(range(10000, 10010)), list(range(-1, -11, -1)), list(range(1000))))
+
+        for args, x1, x2, x3 in cases:
```
Comment (kernc) [inline Orange/tests/test_domain.py]:
```py
for domain_args, attributes, class_vars, metas in (...
```

File: Orange/tests/test_evaluation_scoring.py [hunk 15]
```diff
@@ -62,104 +62,91 @@ def test_bayes(self):
...
+        res = TestOnTrainingData(self.iris, [maj])
         self.assertEqual(AUC(res)[0], 0.5)
 
     def test_multiclass_auc_multi_learners(self):
-        learners = [Orange.classification.LogisticRegressionLearner(),
-                    Orange.classification.MajorityLearner()]
-        res = Orange.evaluation.testing.CrossValidation(self.iris, learners, k=10)
+        learners = [LogisticRegressionLearner(),
+                    MajorityLearner()]
+        res = testing.CrossValidation(self.iris, learners, k=10)
```
Comment (kernc) [inline Orange/tests/test_evaluation_scoring.py]: `CrossValidation` is imported.

File: Orange/tests/test_evaluation_scoring.py [hunk 16]
```diff
@@ -62,104 +62,91 @@ def test_bayes(self):
...
-        # One wrong
-        self.assertAlmostEqual(
-            self.compute_auc(actual, [0., 0., 0., 1., 1., 0.]), 5 / 6)
-        # Two wrong
-        self.assertAlmostEqual(
-            self.compute_auc(actual, [1., 1., 0., 1., 1., 1.]), 4 / 6)
-        # Three wrong
-        self.assertAlmostEqual(
-            self.compute_auc(actual, [1., 1., 0., 1., 1., 0.]), 3 / 6)
+        for predict, prob in (([1., 1., 1., 0., 0., 0.], 0.),      # All wrong
```
Comment (kernc) [inline Orange/tests/test_evaluation_scoring.py]:
```py
for predicted, auc in ...
```

File: Orange/tests/test_evaluation_testing.py [hunk 17]
```diff
@@ -285,19 +282,16 @@ def test_results(self):
         np.testing.assert_equal(res.row_indices, np.arange(nrows))
 
     def test_store_data(self):
-        nrows, ncols = 50, 10
-        t = random_data(nrows, ncols)
         learners = [NaiveBayesLearner()]
-
+        t = random_data(50, 10)
+        learners = learners
```
Comment (kernc) [inline Orange/tests/test_evaluation_testing.py]: What does this line do? :smiley:

File: Orange/tests/test_fss.py [hunk 18]
```diff
@@ -74,27 +74,24 @@ def test_defaults(self):
 
 class TestRemoveNaNColumns(unittest.TestCase):
     def test_column_filtering(self):
-        data = Orange.data.Table("iris")
+        data = Table("iris")
```
Comment (kernc) [inline Orange/tests/test_fss.py]: Could make common datasets global for all the test cases to use and not load separately?

File: Orange/tests/test_fss.py [hunk 19]
```diff
@@ -74,27 +74,24 @@ def test_defaults(self):
...
-        data = Orange.data.Table("voting")
-        srf = Orange.preprocess.SelectRandomFeatures(k=3)
-        new_data = srf(data)
-        self.assertEqual(len(new_data.domain.attributes), 3)
-
-        srf = Orange.preprocess.SelectRandomFeatures(k=0.25)
-        new_data = srf(data)
-        self.assertEqual(len(new_data.domain.attributes), 4)
+        data = Table("voting")
+        for x, y in ((3, 3), (0.25, 4)):
```
Comment (kernc) [inline Orange/tests/test_fss.py]: Nothing wrong if code is more verbose:
```py
for k_features, n_attributes in ...
```

File: Orange/tests/test_impute.py [hunk 20]
```diff
@@ -33,13 +33,11 @@ def test_replacement(self):
...
                              data.ContinuousVariable("c"),)
-        table = data.Table(domain, x, c1)
-        var1 = preprocess.Average()(table, 0)
-        self.assertIsInstance(var1.compute_value, preprocess.ReplaceUnknowns)
-        self.assertEqual(var1.compute_value.value, 0.5)
-        var2 = preprocess.Average()(table, 1)
-        self.assertIsInstance(var2.compute_value, preprocess.ReplaceUnknowns)
-        self.assertEqual(var2.compute_value.value, 2)
+        table = Table(domain, x, c1)
+        for x, y in ((0, 0.5), (1, 2)):
```
Comment (kernc) [inline Orange/tests/test_impute.py]:
```py
for col, computed_value in ...
```

File: Orange/tests/test_impute.py [hunk 21]
```diff
@@ -154,8 +152,8 @@ def test_replacement(self):
 
         v = impute.Model(MajorityLearner())(table, domain[0])
         self.assertTrue(np.all(np.isfinite(v.compute_value(table))))
-        self.assertTrue(np.all(v.compute_value(table) == [1., 2., 1.])
-                        or np.all(v.compute_value(table) == [1., 2., 2.]))
+        self.assertTrue(np.all(v.compute_value(table) == [1., 2., 1.]) or
+                        np.all(v.compute_value(table) == [1., 2., 2.]))
```
Comment (kernc) [inline Orange/tests/test_impute.py]: I don't think I understand this test. Can you explain it?

File: Orange/tests/test_instance.py [hunk 22]
```diff
@@ -124,43 +123,32 @@ def test_init_xy_arr(self):
...
-        self.assertEqual(inst._metas.shape, (3, ))
+        self.assertEqual(inst._x.shape, (2,))
+        self.assertEqual(inst._y.shape, (1,))
+        self.assertEqual(inst._metas.shape, (3,))
         assert_array_equal(inst._x, np.array([42, 0]))
         self.assertEqual(inst._y[0], 1)
         assert_array_equal(inst._metas, np.array([0, 43, "Foo"], dtype=object))
 
+    def test_init_xym_arr(self):
+        self.helper_init_xym_arr(np.array([42, "M", "B", "X", 43, "Foo"], dtype=object))
```
Comment (kernc) [inline Orange/tests/test_instance.py]: Instead of a helper method, you could have single `test_init` and then different `args` applied in a for loop, like with most other tests above.

File: Orange/tests/test_instance.py [hunk 23]
```diff
@@ -191,30 +179,15 @@ def test_get_item(self):
...
-        self.assertEqual(inst[2], "B")
-        self.assertEqual(inst["y"], "B")
-        self.assertEqual(inst[domain.class_var], "B")
-
-        val = inst[-2]
-        self.assertIsInstance(val, Value)
-        self.assertEqual(inst[-2], 43)
-        self.assertEqual(inst["Meta 2"], 43)
-        self.assertEqual(inst[self.metas[1]], 43)
+        for pos, comp, index, dom in ((0, 42, "x", domain[0]),
```
Comment (kernc) [inline Orange/tests/test_instance.py]:
```py
for idx_int, idx_name, idx_var, value in ...
```
Note the reordering.

File: Orange/tests/test_instance.py [hunk 24]
```diff
@@ -241,36 +214,24 @@ def test_set_item(self):
...
-        inst["x"] = 44
-        self.assertEqual(inst[0], 44)
-        inst[domain[0]] = 45
-        self.assertEqual(inst[0], 45)
-
-        inst[1] = "F"
-        self.assertEqual(inst[1], "F")
-        inst["g"] = "M"
-        self.assertEqual(inst[1], "M")
+        for pos, instPos, val in ((0, 0, 42),
```
Comment (kernc) [inline Orange/tests/test_instance.py]:
```py
for idx1, idx2, val in ...
```

File: Orange/tests/test_majority.py [hunk 25]
```diff
@@ -13,55 +13,51 @@ class MajorityTest(unittest.TestCase):
...
-        w = (y == heavy_class) * 2 + 1
+        w = (y == 1) * 2 + 1
         t = Table(x, y, W=w)
-        learn = MajorityLearner()
-        clf = learn(t)
+        clf = self.learn(t)
 
         y2 = clf(x)
-        self.assertTrue((y2 == heavy_class).all())
+        self.assertEqual(y2.all(), 1)
```
Comment (kernc) [inline Orange/tests/test_majority.py]: This test was OK before. `all()` returns a bool which should be true.

File: Orange/tests/test_majority.py [hunk 26]
```diff
@@ -13,55 +13,51 @@ class MajorityTest(unittest.TestCase):
...
         y = clf(iris)
         self.assertTrue((y == 2).all())
 
-        learn = MajorityLearner()
         for e in iris:
             e.set_class("?")
         clf = learn(iris)
         y = clf(iris)
-        self.assertTrue((y == 1).all())
+        self.assertEqual(y.all(), 1)
```
Comment (kernc) [inline Orange/tests/test_majority.py]: As above.

File: Orange/tests/test_majority.py (hunk 25 above)
Comment (kernc) [inline Orange/tests/test_majority.py]: Oh, nevermind, the new test is OK.

File: Orange/tests/test_majority.py [hunk 27]
```diff
@@ -13,55 +13,51 @@ class MajorityTest(unittest.TestCase):
...
+        self.assertEqual(y2.all(), 1)
 
     def test_weights(self):
         nrows = 100
         ncols = 10
         x = np.random.random_integers(1, 3, (nrows, ncols))
         y = np.array(70*[0] + 30*[1]).reshape((nrows, 1))
-        heavy_class = 1
-        w = (y == heavy_class) * 2 + 1
+        w = (y == 1) * 2 + 1
```
Comment (kernc) [inline Orange/tests/test_majority.py]: `heavy_class` was OK. [Prefer symbols to magic numbers](https://en.wikipedia.org/wiki/Magic_number_%28programming%29#Unnamed_numerical_constants).

File: Orange/tests/test_mean.py [hunk 28]
```diff
@@ -28,11 +31,9 @@ def test_weights(self):
         ncols = 10
         x = np.random.random_integers(1, 3, (nrows, ncols))
         y = np.random.random_integers(0, 4, (nrows, 1)) / 3.0
-        heavy = 1
```
Comment (kernc) [inline Orange/tests/test_mean.py]: Leave `heavy` be.

File: Orange/tests/test_naive_bayes.py [hunk 29]
```diff
@@ -3,38 +3,35 @@
...
+from Orange.evaluation import CrossValidation, CA
+
 
 class NaiveBayesTest(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        cls.titanic = Table('titanic')
+        data = Table('titanic')
+        cls.learner = NaiveBayesLearner()
+        cls.learned = cls.learner(data)
```
Comment (kernc) [inline Orange/tests/test_naive_bayes.py]: `cls.model = ...`

File: Orange/tests/test_evaluation_testing.py (hunk 17 above)
Comment (rekonder) [inline Orange/tests/test_evaluation_testing.py]: Nothing smart :). I will delete it

File: Orange/tests/test_domain.py [hunk 30]
```diff
@@ -252,55 +233,29 @@ def test_get_item_slices(self):
...
-        self.assertFalse(race in d)
-        self.assertFalse(3 in d)
-        self.assertFalse(np.int_(3) in d)
-        self.assertFalse(-2 in d)
-        self.assertFalse(np.int_(-2) in d)
+        for var in ["AGE", age, 0, np.int_(0), "income", income, 2, np.int_(2), "SSN", ssn, -1, np.int_(-1)]:
+            self.assertTrue(var in d)
+
+        for var in ["no_such_thing", race, 3, np.int_(3), -2, np.int_(-2)]:
+            self.assertFalse(var in d)
```
Comment (kernc) [inline Orange/tests/test_domain.py]:
```py
self.assertNotIn(var, d)
```

File: Orange/tests/test_domain.py [hunk 31]
```diff
@@ -252,55 +233,29 @@ def test_get_item_slices(self):
...
-        self.assertTrue(np.int_(-1) in d)
-
-        self.assertFalse("no_such_thing" in d)
-        self.assertFalse(race in d)
-        self.assertFalse(3 in d)
-        self.assertFalse(np.int_(3) in d)
-        self.assertFalse(-2 in d)
-        self.assertFalse(np.int_(-2) in d)
+        for var in ["AGE", age, 0, np.int_(0), "income", income, 2, np.int_(2), "SSN", ssn, -1, np.int_(-1)]:
+            self.assertTrue(var in d)
```
Comment (kernc) [inline Orange/tests/test_domain.py]:
```py
self.assertIn(var, d)
```

Comment (rekonder) [inline Orange/tests/test_domain.py]: Changed here and on other places where I find this similar line. Thanks :).

File: Orange/tests/test_impute.py (hunk 21 above)
Comment (rekonder) [inline Orange/tests/test_impute.py]: At the moment not. But I will try figure out.

File: Orange/tests/test_domain.py [hunk 32]
```diff
@@ -340,7 +295,7 @@ def test_has_discrete(self):
         self.assertFalse(Domain([]).has_discrete_attributes())
         self.assertFalse(Domain([], [age]).has_discrete_attributes())
         self.assertFalse(Domain([], race).has_discrete_attributes())
-
+
```
Comment (kernc) [inline Orange/tests/test_domain.py]: This looks like a blank space. You can set PyCharm to strip trailing whitespace on modified lines in Settings > Editor > General.

**Comment by kernc** [comment]
Once you're happy with it, please remove the WIP tag.

**Comment by rekonder** [comment]
I delete one unused import and fix problem with trailing whitespace. Now I will remove [WIP] tag.

File: Orange/tests/test_evaluation_testing.py [hunk 33]
```diff
@@ -381,11 +373,10 @@ def test_probs(self):
         self.assertEqual(results.predicted.shape, (2, len(test)))
         np.testing.assert_equal(results.predicted, np.ones((2, 50)))
         probs = results.probabilities
-        self.assertTrue((probs[:, :, 0] == 0).all())
+        self.assertEqual((probs[:, :, 0]).all(), 0)
```
Comment (kernc) [inline Orange/tests/test_evaluation_testing.py]: Sorry to nitpick here, but please revert this assertion. It was way clearer before. :smiley:

Comment (rekonder) [inline Orange/tests/test_evaluation_testing.py]: Np. I revert this change :)
